### PR TITLE
PR for Trend Micro Vision One connector

### PR DESCRIPTION
**Pull request by alex-y-lin**

### CODE REVIEW

**Comment by delliott90** [comment]
Hi @alex-y-lin, thanks for the submission! Would you be able to sign the CLA? It's a requirement before we would be able to merge this in.

File: stix_shifter_modules/vision_one/stix_transmission/results_connector.py
Comment (mdazam1942) [inline stix_shifter_modules/vision_one/stix_transmission/results_connector.py:15]: is this warning necessary?

File: stix_shifter_modules/vision_one/stix_translation/query_constructor.py
Comment (mdazam1942) [inline stix_shifter_modules/vision_one/stix_translation/query_constructor.py:63]: there are lots of unnecessary `\` in the translated query. you may want to update the method with something like this depending on your native query:
`return '{}'.format(value.replace('\"', '\\"').replace('(', '\\(').replace(')', '\\)'))
`

**Review by mdazam1942 — CHANGES_REQUESTED** [review]
Overall most part of the module looks fine. Need to fix few things before we can merge it. Combined Observation Expression doesn't work for some reason. You may want to have a look. here's an example of a valid stix pattern:
`python main.py translate vision_one query "{}" "[ipv4-addr:value = '66.172.27.225'] OR [ipv6-addr:value = '3001:0:0:0:0:0:0:2']"`

One more thing, I think we should add trendmicro as a prefix in the module name. So the module name can be **trendmicro_vision_one** @delliott90 what do you think?

File: stix_shifter_modules/vision_one/stix_transmission/results_connector.py
Comment (alex-y-lin) [reply stix_shifter_modules/vision_one/stix_transmission/results_connector.py:15]: Users should be informed that if users set length to be greater than 1000, the API will use 1000 instead.

**Comment by alex-y-lin** [comment]
> Overall most part of the module looks fine. Need to fix few things before we can merge it. Combined Observation Expression doesn't work for some reason. You may want to have a look. here's an example of a valid stix pattern and returned error:
> 
> ```
> python main.py translate vision_one query "{}" "[ipv4-addr:value = '66.172.27.225'] OR [ipv6-addr:value = '3001:0:0:0:0:0:0:2']"
> {
>     "success": false,
>     "code": "not_implemented",
>     "error": "wrong parameter : Parenthesis is not support in MDL"
> }
> ```
The dialect, messageActivityData' does not support parenthesis in query.

**Comment by alex-y-lin** [comment]
> Hi @alex-y-lin, thanks for the submission! Would you be able to sign the CLA? It's a requirement before we would be able to merge this in.

I am waiting for my legal team's approval.

File: stix_shifter_modules/vision_one/entry_point.py
Comment (alex-y-lin) [reply stix_shifter_modules/vision_one/entry_point.py:32]: setup_transmission_simple uses `connector = module.StatusConnector(api_client)`. Since there is no api_client argument in constructor of Vision One  StatusConnector, the setup logic can not just use setup_transmission_simple.

File: stix_shifter_modules/vision_one/stix_translation/query_constructor.py
Comment (alex-y-lin) [reply stix_shifter_modules/vision_one/stix_translation/query_constructor.py]: Escaping `\` by `\\` is necessary for native query string of Trend Micro Vision One API.

File: stix_shifter_modules/vision_one/stix_transmission/api_client.py
Comment (mdazam1942) [inline stix_shifter_modules/vision_one/stix_transmission/api_client.py:39]: why are you just returning the response code? PingConnector class calls this method to get the response. response is already handled in `PingConnector.ping_connection()`

File: stix_shifter_modules/vision_one/stix_transmission/results_connector.py
Comment (mdazam1942) [reply stix_shifter_modules/vision_one/stix_transmission/results_connector.py:15]: so does it mean no matter what the user sets the length of the query it will only return 1000 results?

File: stix_shifter_modules/vision_one/configuration/config.json
Comment (mdazam1942) [reply stix_shifter_modules/vision_one/configuration/config.json:41]: still see `tok`. would you please rename it to  `token`

**Comment by mdazam1942** [comment]
> > Overall most part of the module looks fine. Need to fix few things before we can merge it. Combined Observation Expression doesn't work for some reason. You may want to have a look. here's an example of a valid stix pattern and returned error:
> > ```
> > python main.py translate vision_one query "{}" "[ipv4-addr:value = '66.172.27.225'] OR [ipv6-addr:value = '3001:0:0:0:0:0:0:2']"
> > {
> >     "success": false,
> >     "code": "not_implemented",
> >     "error": "wrong parameter : Parenthesis is not support in MDL"
> > }
> > ```
> 
> The dialect, messageActivityData' does not support parenthesis in query.

So this needs to be handled for STIX Combined Observation Expression. that's a valid query I tried. so if messageActivityData doesn't support parenthesis than you need to handle that in query constructor.

File: stix_shifter_modules/vision_one/stix_transmission/results_connector.py
Comment (alex-y-lin) [reply stix_shifter_modules/vision_one/stix_transmission/results_connector.py:15]: If the user sets the length less than or equals to 1000, the query will return the amount of data that the user wants. If the user set the length greater than 1000, the query will return 1000 results only.

File: stix_shifter_modules/vision_one/stix_transmission/api_client.py
Comment (alex-y-lin) [reply stix_shifter_modules/vision_one/stix_transmission/api_client.py:39]: This API is just to check whether the target server is available and token is valid.

**Comment by mdazam1942** [comment]
@alex-y-lin I think the module name should be `trendmicro_vision_one`. its better to have the vendor name as the module name.

**Comment by alex-y-lin** [comment]
> @alex-y-lin I think the module name should be `trendmicro_vision_one`. its better to have the vendor name as the module name.

The module name is renamed. Would you please review the changes?

File: stix_shifter_modules/trendmicro_vision_one/stix_transmission/api_client.py
Comment (mdazam1942) [inline stix_shifter_modules/trendmicro_vision_one/stix_transmission/api_client.py:56]: not sure why you used `while`. try/catch might be more appropriate to gracefully rasise the exceptions.

**Review by mdazam1942 — COMMENTED** [review]
Please go through the comments and let us know if you have any questions and help to fix some issues.

And also would you please add a readme file for your connector module with some details on the api and kind of data that the connector retrieves with some CLI commands. There are some readme examples in most of the module such as https://github.com/opencybersecurityalliance/stix-shifter/blob/develop/stix_shifter_modules/arcsight/stix_translation/README.md

File: stix_shifter_modules/trendmicro_vision_one/stix_transmission/api_client.py
Comment (alex-y-lin) [reply stix_shifter_modules/trendmicro_vision_one/stix_transmission/api_client.py]: I use for loop instead. The max fetch count for self-protection. If reaching the max fetch count but not retrieve all data yet, just add log for warning and return the fetched data.

File: stix_shifter_modules/trendmicro_vision_one/configuration/lang_en.json
Comment (mdazam1942) [inline stix_shifter_modules/trendmicro_vision_one/configuration/lang_en.json:16]: `Trend Micro Connection Certificate`

File: stix_shifter_modules/trendmicro_vision_one/stix_translation/query_constructor.py
Comment (alex-y-lin) [reply stix_shifter_modules/trendmicro_vision_one/stix_translation/query_constructor.py]: I have implemented the limitation of messageActivityData query. The messageActivityData dialect does not support: 1. using AND and OR simultaneously. 2. using multiple criteria for the same field. 3. using NOT.

**Comment by mdazam1942** [comment]
> Please go through the comments and let us know if you have any questions and help to fix some issues.
> 
> And also would you please add a readme file for your connector module with some details on the api and kind of data that the connector retrieves with some CLI commands. There are some readme examples in most of the module such as https://github.com/opencybersecurityalliance/stix-shifter/blob/develop/stix_shifter_modules/arcsight/stix_translation/README.md

@alex-y-lin can you please add a readme.md for your module? details in the comment ^^

**Review by mdazam1942 — APPROVED** [review]
Thanks for making all the changes. Looks good now.

**Pull request merged**
